src/pybind/mgr/cephadm/exchange.py

```diff
@@ -0,0 +1,164 @@
+# Data exchange formats for communicating more
+# complex data structures between the cephadm binary
+# an the mgr module.
+
+import json
+
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    TypeVar,
+    Union,
+    cast,
+)
+
+
+FuncT = TypeVar("FuncT", bound=Callable)
+
+
+class _DataField:
+    """A descriptor to map object fields into a data dictionary."""
+
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        field_type: Optional[FuncT] = None,
+    ):
+        self.name = name
+        self.field_type = field_type
+
+    def __set_name__(self, _: str, name: str) -> None:
+        if not self.name:
+            self.name = name
+
+    def __get__(self, obj: Any, objtype: Any = None) -> Any:
+        return obj.data[self.name]
+
+    def __set__(self, obj: Any, value: Any) -> None:
+        if self.field_type is not None:
+            obj.data[self.name] = self.field_type(value)
+        else:
+            obj.data[self.name] = value
```

Comment:
Trying to get a clear idea of what these repesent. so for
```
fsid = _DataField(field_type=str)
````
if I do a 
```
fsid = 'xxx'
```
does that run through the `__set__` function? and for
```
x = fsid
```
would that use `__get__`

Reply:
Yes, but to be clear these descriptors require a "parent" type. So
```python
t = MyType()
t.fsid = 'abc'  # goes through <Descriptor>.__set__
# and
x = t.fsid  # goes through <Descriptor>.__get__
```